File: tom/tom.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include <errno.h>
#include <stdbool.h>

#define PI 3.14159265358979323846
#define EPS 0.0000001

bool is_around(float num) {
    return num < EPS && num > -EPS;
}

int main() {
    int len = 0;
    scanf("%d", &len);
    int in[5000] = {0};

    int i = 0;
    while (scanf("%d", &in[i++]) == 1);

    int ac = 0;
    int ra = 0;
    int ob = 0;
    for(int i = 0; i < len; i++){
        for(int j = i + 1; j < len; j++){
            for(int k = j + 1; k < len; k++){
                int a = in[i];
                int b = in[j];
                int c = in[k];
                float s = (a + b - c) * (a + c - b) * (c + b - a);
                if (s < EPS)
                    continue;

                float A, B, C;

                errno = 0;
                C = acos((pow(c,2) - pow(a, 2) - pow(b, 2))/(-2.0*a*b));
                if (C < EPS || errno || isnan(C))
                    continue;
                errno = 0;
                B = asin((b*sin(C))/c);
                if (B < EPS || errno || isnan(B))
                    continue;
                A = PI - B - C;
                if (A < EPS)
                    continue;

                if (fabs(A - (PI / 2)) < EPS || fabs(B - (PI / 2)) < EPS || fabs(C - (PI / 2)) < EPS){
                    ra++;
                } else if (A > (PI / 2) || B > (PI / 2) || C > (PI / 2)) {
                    ob++;
                } else {
                    ac++;
                }
            }
        }
    }
    printf("%i %i %i\n", ac, ra, ob);

    return 0;
}

File: tom/tom.py
import math
import threading

nl = int(input())
sl = [int(i) for i in input().split(" ")]
ac = 0
ra = 0
ob = 0
eps = 0.0000001
num_threads = 4


class TriangleThread(threading.Thread):
    def __init__(self, triangles):
        threading.Thread.__init__(self)
        self.triangles = triangles
        self.ra = 0
        self.ob = 0
        self.ac = 0

    def run(self):
        for t in self.triangles:
            a, b, c = t[0], t[1], t[2]
            s = (a + b - c) * (a + c - b) * (c + b - a)
            if s < eps:
                continue
            try:
                C = math.acos((c ** 2 - a ** 2 - b ** 2) / (-2 * a * b))
                B = math.asin((b * math.sin(C)) / c)
                A = math.pi - B - C
                if math.pi / 2 in [A, B, C]:
                    self.ra += 1
                elif max([A, B, C]) > math.pi / 2:
                    self.ob += 1
                else:
                    self.ac += 1
            except:
                continue


all_permuations = []
for i in range(len(sl)):
    for j in range(i + 1, len(sl)):
        for k in range(j + 1, len(sl)):
            all_permuations.append([sl[i], sl[j], sl[k]])
threads = []
chunks = [all_permuations[x:x+num_threads] for x in range(0, len(all_permuations), num_threads)]
for chunk in chunks:
    threads.append(TriangleThread(chunk))

for T in threads:
    T.start()

for T in threads:
    T.join()

print(sum([T.ac for T in threads]), sum([T.ra for T in threads]), sum([T.ob for T in threads]))
